File: tier2_intermediate/dynamic_programming/problems/p467_unique_substrings_in_wraparound_string.rs
/*
 * P467: Unique Substrings in Wraparound String [PREMIUM] (Medium)
 * https://leetcode.com/problems/unique-substrings-in-wraparound-string/
 * Topics: String, Dynamic Programming
 * 
 * We define the string base to be the infinite wraparound string of "abcdefghijklmnopqrstuvwxyz", so base will look like this:
 * Given a string s, return the number of unique non-empty substrings of s are present in base.
 * Example 1:
 *     Input: s = "a"
 *     Output: 1
 *     Explanation: Only the substring "a" of s is in base.
 * 
 * Example 2:
 *     Input: s = "cac"
 *     Output: 2
 *     Explanation: There are two substrings ("a", "c") of s in base.
 * 
 * Example 3:
 *     Input: s = "zab"
 *     Output: 6
 *     Explanation: There are six substrings ("z", "a", "b", "za", "ab", and "zab") of s in base.
 * 
 * Constraints:
 *     - 1 <= s.length <= 105
 *     - s consists of lowercase English letters.
 * 
 * Hint: One possible solution might be to consider allocating an array size of 26 for each character in the alphabet. (Credits to @r2ysxu)
 * 
 * Template (python3):
 *     class Solution:
 *         def findSubstringInWraproundString(self, s: str) -> int:
 */
use rstest;

fn find_substring_in_wrapround_string(s: &str) -> i32 {
    todo!()
}

fn main() {
    struct TC<'a> { label: &'a str, s: &'a str, expected: i32 }
    let tests: &[TC] = &[
        TC { label: "example 1", s: "a", expected: 1 },
        TC { label: "example 2", s: "cac", expected: 2 },
        TC { label: "example 3", s: "zab", expected: 6 },
        TC { label: "all same chars only a in wraparound", s: "aaaaa", expected: 1 },
        TC { label: "wraparound chain length 8", s: "zabcdefg", expected: 36 },
        TC { label: "full alphabet", s: "abcdefghijklmnopqrstuvwxyz", expected: 351 },
    ];
    let mut passed = 0;
    for (i, tc) in tests.iter().enumerate() {
        let got = find_substring_in_wrapround_string(tc.s);
        if got == tc.expected {
            passed += 1;
            println!("  Test {} ({}): PASS", i + 1, tc.label);
        } else {
            println!("  Test {} ({}): FAIL (expected {}, got {})", i + 1, tc.label, tc.expected, got);
        }
    }
    println!("\n  {}/{} passed", passed, tests.len());
    std::process::exit(if passed == tests.len() { 0 } else { 1 });
}

File: tier2_intermediate/dynamic_programming/test_runners/p467_unique_substrings_in_wraparound_string.py
from src.utils.judge_base import JudgeBase, TestCase, register


class P467UniqueSubstringsWraparound(JudgeBase):
    name = "467. Unique Substrings in Wraparound String"
    test_cases = [
        TestCase(input="a", expected=1, label="example 1"),
        TestCase(input="cac", expected=2, label="example 2"),
        TestCase(input="zab", expected=6, label="example 3"),
        TestCase(input="aaaaa", expected=1, label="all same chars only a in wraparound"),
        TestCase(input="zabcdefg", expected=36, label="wraparound chain length 8"),
        TestCase(input="abcdefghijklmnopqrstuvwxyz", expected=351, label="full alphabet"),
    ]

    def to_stdin(self, tc_input) -> str:
        return self.serialize_string(tc_input)

    def check_stdout(self, stdout: str, tc_expected) -> bool:
        return self.parse_int(stdout) == tc_expected


register(P467UniqueSubstringsWraparound)
